pkg/utils/Idpool.go

```diff
@@ -0,0 +1,154 @@
+#   ID so that a caller knows when an ID was newly assigned (ref_count 1) or
+#   finally released (ref_count 0).
+# ---------------------------------------------------------------------------------
+*/
+type IDPool struct {
+	// self._lock = threading.Lock()
+	name         string                 // Name of pool
+	_unusedIDs   []uint32               // Yet unused IDs in pool Available ids
+	_idsInUse    map[interface{}]uint32 // Mapping key: id for currently assigned ids
+	_idsForReuse map[interface{}]uint32 // Mapping key: id for previously assigned ids
+	_refs        map[uint32][]interface{}
+	_size        int // Size of the pool
+}
+
+// IDPoolInit initialize mod ptr pool
+func IDPoolInit(name string, min uint32, max uint32) IDPool {
+	var id IDPool
+	id.name = name
+	id._unusedIDs = make([]uint32, 0)
+	for j := min; j <= (max + 1); j++ {
+		id._unusedIDs = append(id._unusedIDs, j)
+	}
+	id._size = len(id._unusedIDs)
+	id._idsInUse = make(map[interface{}]uint32)
+	id._idsForReuse = make(map[interface{}]uint32)
+	id._refs = make(map[uint32][]interface{})
+	return id
+}
+
+func (ip *IDPool) _assignID(key interface{}) uint32 {
+	// Check if there was an id assigned for that key earlier
+	id := ip._idsForReuse[key]
+	if !reflect.ValueOf(id).IsZero() {
+		// Re-use the old id
+		delete(ip._idsForReuse, key)
+	} else {
+		if len(ip._unusedIDs) != 0 {
+			// Pick an unused id
+			id = ip._unusedIDs[0]
+			ip._unusedIDs = append(ip._unusedIDs[:0], ip._unusedIDs[1:]...)
+		} else {
+			if len(ip._idsForReuse) != 0 {
+				// Pick one of the ids earlier used for another key
+				for oldKey := range ip._idsForReuse {
+					delete(ip._idsForReuse, oldKey)
+					break
+				}
+			} else {
+				// No id left
+				log.Printf("IDPool: Failed to allocate id for %+v. No free ids in pool.", key)
+				return 0
+			}
+		}
+	}
+	// Store the assigned id, if any
+	if !reflect.ValueOf(id).IsZero() {
+		ip._idsInUse[key] = id
+	}
+	return id
+}
+
+// GetID get the mod ptr id from pool
+func (ip *IDPool) GetID(key interface{}, ref interface{}) (uint32, uint32) {
+	id := ip._idsInUse[key]
+	if reflect.ValueOf(id).IsZero() {
+		// Assign a free id for the key
+		id = ip._assignID(key)
+		if id == 0 {
+			return 0, 0
+		}
+	}
+	if !reflect.ValueOf(ref).IsZero() {
+		log.Printf("IDPool: GetID  Assigning key : %+v , id  %+v for ref %v", id, key, ref)
+		// refSet := ip._refs[id]
+		if reflect.ValueOf(ip._refs[id]).IsZero() {
+			ip._refs[id] = make([]interface{}, 0)
+		}
+		ip._refs[id] = append(ip._refs[id], ref)
+		return id, uint32(len(ip._refs[id]))
+	}
+	log.Printf("IDPool: GetID Assigning id %v for key %v and ref %v", id, key, ref)
+	return id, uint32(0)
+}
+
+func deleteRef(refSet []interface{}, ref interface{}) []interface{} {
+	// size := len(refSet)
+	var i uint32
+	for index, value := range refSet {
+		if value == ref {
+			i = uint32(index)
+			break
+		}
+	}
+	return append(refSet[:i], refSet[i+1:]...)
```

Comment:
if you need to perform such operations, why not use `map[key]struct{}`?

Reply:
It is for deleting a ref element from the refset(array) pointing to an id index.
we feel array is preferable instead of struct

Reply:
still do not understand why we cannot use `map[key]struct{}` instead of `refSet []interface{}`

Reply:
actually we need golang type equivalent of 'list' type in python.  
we have replaced with "ref map[uint32]map[interface{}]bool"
here the scenario is for same key we can have different references of interface type as reference can be of any type like int, string, alphanumeric etc. that's why we cannot use counter here.